File: src/core/parser.ts
/**
 * @file parser.ts
 * @description 解析器模块，用于解析页面内容和提取所需信息。
 * @module parser
 * @requires cheerio - 用于解析HTML的库。
 * @exports parsePageLinks - 解析页面中的电影链接。
 * @exports parseMetadata - 解析页面中的元数据。
 * @exports parseCategories - 解析页面中的分类信息。
 * @exports parseActress - 解析页面中的演员信息。
 * @exports parseFilmData - 解析页面中的影片数据。
 * @author raawaa
 */

import { Config, Metadata, FilmData } from '../types/interfaces';
import logger from './logger';

class Parser {
  private config: Config;

  public constructor(config: Config) {
    this.config = config;
  }

  /**
   * 解析页面中的电影链接
   * @param {string} html - 页面HTML内容
   * @returns {Array<string>} 电影详情页链接数组
   */
  static parsePageLinks(html: string): Array<string> {
    // 检查页面内容是否为空
    if (!html || html.length === 0) {
      logger.warn('parsePageLinks: 接收到空的HTML内容');
      return [];
    }

    const $ = require('cheerio').load(html);
    
    const links: string[] = Array.from(
      new Set(
        $('a.movie-box')
          .map((i: number, el: cheerio.Element) => $(el).attr('href'))
          .get()
          .map((link: unknown) => String(link || '').trim())
          .filter(Boolean)
      )
    );
    
    logger.debug(`解析到 ${links.length} 个影片链接`);
    if (links.length === 0) {
      // 只在调试模式下输出完整HTML或其片段，避免日志过长
      logger.debug('页面中未找到影片链接，页面内容片段 (前1000字符):');
      logger.debug(html.substring(0, 1000));
    }
    
    return links;
  }


  /**
   * 解析页面中的元数据
   * @param {string} html - 页面 HTML 内容
   * @returns {Metadata} 包含影片元数据的对象
   * @throws {Error} 当无法从脚本中解析出所需元数据时抛出错误
   * @description 从页面 HTML 内容中提取影片的 gid、uc、img、标题、分类和演员信息
   */
  static parseMetadata(html: string) {
    if (!html || typeof html !== 'string') {
      logger.warn('parseMetadata: HTML内容为空或不是字符串');
      throw new Error('Invalid HTML content for metadata parsing');
    }

    if (html.length === 0) {
      logger.warn('parseMetadata: HTML内容长度为0');
      throw new Error('Empty HTML content for metadata parsing');
    }

    const $ = require('cheerio').load(html);
    const scripts = $('script', 'body')
      .map((_: number, el: cheerio.Element) => $(el).html() || '')
      .get()
      .filter(Boolean);

    logger.debug(`parseMetadata: 页面中找到 ${scripts.length} 个script标签`);

    const primaryScript = scripts.find((item: string) =>
      /(?:\bgid\b|\buc\b|\bimg\b|uncledatoolsbyajax|sample_dmm)/i.test(item)
    );
    const metadataSources = [primaryScript || '', ...scripts, html].filter(Boolean);

    const gid = this.extractMetadataField(metadataSources, [
      /\bgid\b\s*[:=]\s*['"]?(\d+)['"]?/i,
      /["']gid["']\s*:\s*["']?(\d+)["']?/i,
      /\bgid\b\s*[:=]\s*parseInt\(\s*['"]?(\d+)['"]?/i,
      /[?&]gid=(\d+)/i,
      /data-gid\s*=\s*["'](\d+)["']/i
    ]);
    const uc = this.extractMetadataField(metadataSources, [
      /\buc\b\s*[:=]\s*['"]?(\d+)['"]?/i,
      /["']uc["']\s*:\s*["']?(\d+)["']?/i,
      /\buc\b\s*[:=]\s*parseInt\(\s*['"]?(\d+)['"]?/i,
      /[?&]uc=(\d+)/i,
      /data-uc\s*=\s*["'](\d+)["']/i
    ]);
    const img =
      this.extractMetadataField(metadataSources, [
        /\bimg\b\s*[:=]\s*['"]([^'"]+)['"]/i,
        /["']img["']\s*:\s*["']([^"']+)["']/i,
        /[?&]img=([^&"'>\s]+)/i,
        /data-img\s*=\s*["']([^"']+)["']/i
      ])?.replace(/\\\//g, '/') ||
      $('.bigImage img').attr('src') ||
      $('a.bigImage img').attr('src') ||
      $('meta[property="og:image"]').attr('content') ||
      '';

    logger.debug(`解析脚本内容: gidMatch=${Boolean(gid)}, ucMatch=${Boolean(uc)}, imgMatch=${Boolean(img)}`);
    logger.debug(`主脚本长度: ${(primaryScript || '').length} 字符`);

    if (!gid || !uc || !img) {
      logger.warn('parseMetadata: 无法从页面中解析出完整元数据');
      logger.debug(`主脚本内容: ${(primaryScript || '').substring(0, 500)}`);
      logger.debug(`尝试替代解析: altGidMatch=${Boolean(gid)}, altUcMatch=${uc || null}`);
      throw new Error('Failed to parse required metadata from page');
    }
    
    const metadata: Metadata = {
      gid,
      uc,
      img,
      title:
        $('h3').first().text().trim() ||
        $('meta[property="og:title"]').attr('content')?.trim() ||
        $('title').text().trim(),
      category: this.parseCategories($),
      actress: this.parseActress($)
    };
    
    logger.debug(`解析到影片元数据: 标题=${metadata.title}, gid=${metadata.gid}, uc=${metadata.uc}`);
    return metadata;
  }

  private static extractMetadataField(sources: string[], patterns: RegExp[]): string | null {
    for (const source of sources) {
      for (const pattern of patterns) {
        const match = pattern.exec(source);
        if (match?.[1]) {
          return match[1].trim();
        }
      }
    }

    return null;
  }


  /**
   * 解析HTML内容中的影片分类信息
   * @param {any} $ - Cheerio对象或包含分类信息的HTML字符串
   * @returns {Array<string>} 返回分类名称的数组
   * @description 从HTML中提取所有位于<span class="genre">标签内，
   * 且嵌套在<label><a>结构中的文本内容作为分类名称
   */
  static parseCategories($: any): Array<string> {
    // 如果传入的是HTML字符串，则加载为Cheerio对象
    if (typeof $ === 'string') {
      if (!$ || $.length === 0) {
        logger.warn('parseCategories: HTML内容为空或不是字符串');
        return [];
      }
      $ = require('cheerio').load($);
    } else if (!$ || typeof $ !== 'function') {
      logger.warn('parseCategories: 传入的不是有效的Cheerio对象或HTML字符串');
      return [];
    }

    // 检查页面是否包含关键HTML结构
    const hasGenreElements = $('span.genre').length > 0;
    const hasLabelElements = $('span.genre label').length > 0;
    const hasAnchorElements = $('span.genre a').length > 0;

    if (!hasGenreElements) {
      logger.debug('parseCategories: 页面中没有找到 span.genre 元素');
      logger.debug(`页面片段: ${$.html ? $.html().substring(0, 500) : '无法获取HTML内容'}`);
    }

    if (!hasLabelElements && hasGenreElements) {
      logger.debug('parseCategories: 找到 span.genre 但没有找到 label 元素');
    }

    if (!hasAnchorElements && hasLabelElements) {
      logger.debug('parseCategories: 找到 label 但没有找到 a 元素');
    }

    const categories = $('span.genre label a').map((i: number, el: cheerio.Element) => $(el).text()).get();
    logger.debug(`解析到 ${categories.length} 个分类: ${categories.join(', ')}`);

    if (categories.length === 0 && hasGenreElements) {
      logger.debug('parseCategories: 页面包含genre元素但未能解析到分类文本');
      // 输出一些元素用于调试
      const sampleGenre = $('span.genre').first();
      if (sampleGenre.length > 0) {
        logger.debug(`第一个genre元素HTML: ${sampleGenre.html()}`);
      }
    }

    return categories;
  }


  /**
   * 解析HTML内容中的女演员信息
   * @param {any} $ - Cheerio对象或包含演员信息的HTML字符串
   * @returns {Array<string>} 返回女演员名称的数组
   * @description 从HTML中提取所有位于.star-name .a标签内的文本内容作为女演员名称
   */
  static parseActress($: any): Array<string> {
    // 如果传入的是HTML字符串，则加载为Cheerio对象
    if (typeof $ === 'string') {
      if (!$ || $.length === 0) {
        logger.warn('parseActress: HTML内容为空或不是字符串');
        return [];
      }
      $ = require('cheerio').load($);
    } else if (!$ || typeof $ !== 'function') {
      logger.warn('parseActress: 传入的不是有效的Cheerio对象或HTML字符串');
      return [];
    }

    // 检查页面是否包含关键HTML结构
    const hasStarNameElements = $('.star-name').length > 0;
    const hasStarBoxElements = $('.star-box').length > 0;
    const hasAnchorElements = $('.star-name a').length > 0;
    const hasActressSection = $.html ? ($.html().includes('演員') || $.html().includes('女優')) : false;

    if (!hasActressSection) {
      logger.debug('parseActress: 页面中没有找到演员相关文字');
    }

    if (!hasStarNameElements && hasStarBoxElements) {
      logger.debug('parseActress: 找到 star-box 但没有找到 star-name 元素');
    }

    if (!hasAnchorElements && hasStarNameElements) {
      logger.debug('parseActress: 找到 star-name 但没有找到 a 元素');
    }

    const actresses = $('.star-name a').map((i: number, el: cheerio.Element) => $(el).text()).get();
    logger.debug(`解析到 ${actresses.length} 个演员: ${actresses.join(', ')}`);

    if (actresses.length === 0 && hasStarBoxElements) {
      logger.debug('parseActress: 页面包含star-box元素但未能解析到演员信息');
      // 输出一些元素用于调试
      const sampleStarBox = $('.star-box').first();
      if (sampleStarBox.length > 0) {
        logger.debug(`第一个star-box元素HTML: ${sampleStarBox.html()}`);
      }
    }

    return actresses;
  }

  /**
   * 将解析的元数据和磁力链接组合成影片数据对象
   * @param {Metadata} metadata - 包含影片元数据的对象
   * @param {string} magnet - 影片的磁力链接，如果没有则为null
   * @param {string} link - 影片详情页链接
   * @returns {FilmData} 返回包含完整影片数据的对象
   * @description 将影片标题、分类、演员信息和磁力链接组合成一个完整的数据对象，
   * 便于后续处理和存储
   */
  static parseFilmData(metadata: Metadata, link: string): FilmData {
    const actressCount = Array.isArray(metadata.actress) ? metadata.actress.length : 0;
    const filmData: FilmData = {
      title: metadata.title,
      sourceLink: link,
      category: metadata.category,
      actress: metadata.actress,
      actressCount
    };

    if (typeof metadata.img === 'string' && metadata.img.trim()) {
      filmData.coverImage = metadata.img.trim();
    }

    return filmData;
  }

  /**
   * 提取防屏蔽地址
   * @param {string} html - 页面HTML内容
   * @returns {Array<string>} 防屏蔽地址数组
   * @description 从页面中提取所有包含防屏蔽地址的链接
   */
  static extractAntiBlockUrls(html: string): Array<string> {
    const $ = require('cheerio').load(html);
    const antiBlockUrls: Array<string> = [];

    // 定位到包含防屏蔽地址的警告框
    const alertBox = $('.alert.alert-info');
    
    // 遍历所有包含防屏蔽地址的列
    alertBox.find('.col-xs-12.col-md-6.col-lg-3.text-center').each((i: number, elem: cheerio.Element) => {
        const strongText = $(elem).find('strong').text().trim();
        
        // 验证是否是防屏蔽地址条目
        if (strongText.includes('防屏蔽地址')) {
            const url = $(elem).find('a').attr('href').trim();
            antiBlockUrls.push(url);
        }
    });

    return antiBlockUrls;
}

}

export default Parser;

File: wails-shell/internal/crawlparse/parser.go
package crawlparse

import (
	"fmt"
	"net/url"
	"regexp"
	"strings"

	"golang.org/x/net/html"
)

// Package crawlparse owns the HTML-to-structured-data extraction layer for
// index/detail/magnet pages. If the site markup changes, start here.
//
// Ownership summary:
// 1) parse raw HTML into crawl metadata/index/magnet structures
// 2) keep selector/pattern heuristics local to one extraction layer
// 3) avoid mixing network retry or output policy into parsing code
//
// File map for maintainers:
// 1) metadata/index/magnet parse contracts
// 2) regex/selector heuristics for site markup extraction
// 3) HTML traversal and structured result assembly helpers

var (
	gidPatterns = []*regexp.Regexp{
		regexp.MustCompile(`(?i)\bgid\b\s*[:=]\s*['"]?(\d+)['"]?`),
		regexp.MustCompile(`(?i)["']gid["']\s*:\s*["']?(\d+)["']?`),
		regexp.MustCompile(`(?i)\bgid\b\s*[:=]\s*parseInt\(\s*['"]?(\d+)['"]?`),
		regexp.MustCompile(`(?i)[?&]gid=(\d+)`),
		regexp.MustCompile(`(?i)data-gid\s*=\s*["'](\d+)["']`),
	}
	ucPatterns = []*regexp.Regexp{
		regexp.MustCompile(`(?i)\buc\b\s*[:=]\s*['"]?(\d+)['"]?`),
		regexp.MustCompile(`(?i)["']uc["']\s*:\s*["']?(\d+)["']?`),
		regexp.MustCompile(`(?i)\buc\b\s*[:=]\s*parseInt\(\s*['"]?(\d+)['"]?`),
		regexp.MustCompile(`(?i)[?&]uc=(\d+)`),
		regexp.MustCompile(`(?i)data-uc\s*=\s*["'](\d+)["']`),
	}
	imgPatterns = []*regexp.Regexp{
		regexp.MustCompile(`(?i)\bimg\b\s*[:=]\s*['"]([^'"]+)['"]`),
		regexp.MustCompile(`(?i)["']img["']\s*:\s*["']([^"']+)["']`),
		regexp.MustCompile(`(?i)[?&]img=([^&"'>\s]+)`),
		regexp.MustCompile(`(?i)data-img\s*=\s*["']([^"']+)["']`),
	}
)

// Metadata is the detail-page parse contract used before magnet fetching and
// filmData projection.
type Metadata struct {
	GID         string   `json:"gid"`
	UC          string   `json:"uc"`
	Img         string   `json:"img"`
	Title       string   `json:"title"`
	Category    []string `json:"category"`
	Actress     []string `json:"actress"`
	ReleaseDate string   `json:"releaseDate,omitempty"`
	Maker       string   `json:"maker,omitempty"`
	Label       string   `json:"label,omitempty"`
	Series      string   `json:"series,omitempty"`
}

// FilmData is the persisted crawler artifact shape written to filmData.json.
type FilmData struct {
	Title       string   `json:"title"`
	SourceLink  string   `json:"sourceLink"`
	Category    []string `json:"category"`
	Actress     []string `json:"actress"`
	CoverImage  string   `json:"coverImage,omitempty"`
	ReleaseDate string   `json:"releaseDate,omitempty"`
	Maker       string   `json:"maker,omitempty"`
	Label       string   `json:"label,omitempty"`
	Series      string   `json:"series,omitempty"`
}

// ParsePageLinks extracts movie-box links from one listing page and leaves
// paging/control flow to upper layers.
func ParsePageLinks(htmlText string) []string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return nil
	}
	links := make([]string, 0)
	seen := map[string]struct{}{}
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "a" && (hasClass(node, "movie-box") || looksLikeFilmDetailHref(getAttr(node, "href"))) {
			if href := strings.TrimSpace(getAttr(node, "href")); href != "" {
				if _, exists := seen[href]; !exists {
					seen[href] = struct{}{}
					links = append(links, href)
				}
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return links
}

func looksLikeFilmDetailHref(href string) bool {
	trimmed := strings.TrimSpace(href)
	if trimmed == "" || strings.HasPrefix(trimmed, "#") || strings.HasPrefix(strings.ToLower(trimmed), "javascript:") {
		return false
	}
	parsed, err := url.Parse(trimmed)
	if err == nil {
		trimmed = parsed.Path
	}
	segment := strings.Trim(strings.TrimSpace(trimmed), "/")
	if segment == "" || strings.Contains(segment, "/") {
		return false
	}
	return regexp.MustCompile(`(?i)^[a-z]{2,12}-?\d{1,8}[a-z]*$`).MatchString(segment)
}

// ParseMetadata converts one detail page into the structured fields required by
// magnet fetching and artifact writing.
func ParseMetadata(htmlText string) (Metadata, error) {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return Metadata{}, err
	}

	scriptTexts := collectElementTexts(root, "script")
	primarySources := make([]string, 0, len(scriptTexts)+1)
	for _, scriptText := range scriptTexts {
		if containsAnyIgnoreCase(scriptText, []string{"gid", "uc", "img", "uncledatoolsbyajax", "sample_dmm"}) {
			primarySources = append(primarySources, scriptText)
		}
	}
	sources := append(primarySources, scriptTexts...)
	sources = append(sources, htmlText)

	gid := extractMetadataField(sources, gidPatterns)
	uc := extractMetadataField(sources, ucPatterns)
	img := strings.ReplaceAll(extractMetadataField(sources, imgPatterns), `\/`, `/`)
	if img == "" {
		img = firstNonEmpty(
			findImageByAncestorClass(root, "div", "bigImage"),
			findImageByAncestorClass(root, "a", "bigImage"),
			findMetaContentByProperty(root, "og:image"),
		)
	}

	if gid == "" || uc == "" || img == "" {
		return Metadata{}, fmt.Errorf("failed to parse required metadata from page")
	}

	title := firstNonEmpty(
		findFirstElementText(root, "h3"),
		findMetaContentByProperty(root, "og:title"),
		findFirstElementText(root, "title"),
	)

	return Metadata{
		GID:         gid,
		UC:          uc,
		Img:         img,
		Title:       title,
		Category:    ParseCategories(htmlText),
		Actress:     ParseActress(htmlText),
		ReleaseDate: ParseReleaseDate(htmlText),
		Maker:       ParseInfoField(htmlText, infoFieldMaker),
		Label:       ParseInfoField(htmlText, infoFieldLabel),
		Series:      ParseInfoField(htmlText, infoFieldSeries),
	}, nil
}

// Info field labels vary with the page locale and have changed slightly over
// time. Keep the accepted aliases in one place so a markup/translation change
// does not leak into the crawler runner or output writer.
const (
	infoFieldMaker  = "maker"
	infoFieldLabel  = "label"
	infoFieldSeries = "series"
)

var infoFieldAliases = map[string][]string{
	infoFieldMaker: {
		"maker", "manufacturer", "studio", "productioncompany", "producer",
		"メーカー", "制作商", "制作公司", "片商",
	},
	infoFieldLabel: {
		"label", "レーベル", "厂牌",
	},
	infoFieldSeries: {
		"series", "シリーズ", "系列",
	},
}

// ParseInfoField extracts one value from JavBus-style information rows. The
// parser intentionally uses the row's label and descendant anchor rather than
// positional selectors, because rows are optional and their order is not
// stable across mirrors/locales.
func ParseInfoField(htmlText string, field string) string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return ""
	}
	aliases := infoFieldAliases[strings.ToLower(strings.TrimSpace(field))]
	if len(aliases) == 0 {
		return ""
	}

	var result string
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != "" || node == nil {
			return
		}
		if node.Type == html.ElementNode && (node.Data == "p" || node.Data == "li" || node.Data == "div") {
			labelNode := firstDescendantByTag(node, "span")
			label := normalizeInfoFieldLabel(nodeText(labelNode))
			if matchesInfoFieldAlias(label, aliases) {
				result = firstNonEmpty(
					firstDescendantTextByTag(node, "a"),
					firstDescendantTextByTag(node, "strong"),
					valueTextAfterLabel(node, labelNode),
				)
				return
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return strings.TrimSpace(result)
}

func normalizeInfoFieldLabel(value string) string {
	value = strings.TrimSpace(strings.ToLower(value))
	return strings.Map(func(r rune) rune {
		switch r {
		case ':', '\uff1a', ' ', '\t', '\r', '\n':
			return -1
		default:
			return r
		}
	}, value)
}

func matchesInfoFieldAlias(label string, aliases []string) bool {
	if label == "" {
		return false
	}
	for _, alias := range aliases {
		if label == normalizeInfoFieldLabel(alias) {
			return true
		}
	}
	return false
}

func firstDescendantByTag(root *html.Node, tag string) *html.Node {
	if root == nil {
		return nil
	}
	var result *html.Node
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != nil || node == nil {
			return
		}
		if node.Type == html.ElementNode && strings.EqualFold(node.Data, tag) {
			result = node
			return
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	for child := root.FirstChild; child != nil; child = child.NextSibling {
		walk(child)
	}
	return result
}

func firstDescendantTextByTag(root *html.Node, tag string) string {
	return strings.TrimSpace(nodeText(firstDescendantByTag(root, tag)))
}

func valueTextAfterLabel(root, labelNode *html.Node) string {
	if root == nil {
		return ""
	}
	text := strings.TrimSpace(nodeText(root))
	label := strings.TrimSpace(nodeText(labelNode))
	if label == "" {
		return text
	}
	text = strings.TrimSpace(strings.TrimPrefix(text, label))
	return strings.TrimSpace(strings.TrimLeft(text, ":： \\t\\r\\n"))
}

// ParseCategories extracts normalized genre labels from one detail page.
func ParseCategories(htmlText string) []string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return nil
	}
	values := make([]string, 0)
	seen := map[string]struct{}{}
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "a" && hasAncestor(node, func(parent *html.Node) bool {
			return parent.Data == "span" && hasClass(parent, "genre")
		}) {
			value := strings.TrimSpace(nodeText(node))
			if value != "" {
				if _, exists := seen[value]; !exists {
					seen[value] = struct{}{}
					values = append(values, value)
				}
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return values
}

// ParseActress extracts the actress list exactly as the detail page exposes it.
func ParseActress(htmlText string) []string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return nil
	}
	values := make([]string, 0)
	seen := map[string]struct{}{}
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "a" && hasAncestor(node, func(parent *html.Node) bool {
			return hasClass(parent, "star-name")
		}) {
			value := strings.TrimSpace(nodeText(node))
			if value != "" {
				if _, exists := seen[value]; !exists {
					seen[value] = struct{}{}
					values = append(values, value)
				}
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return values
}

// releaseDatePattern normalizes recognized date strings into YYYY-MM-DD.
// It tolerates common JAV site formats: "2023-01-01", "2023/01/01",
// "2023年01月01日", and "Jan 01, 2023".
var releaseDatePattern = regexp.MustCompile(`(?i)(\d{4})\s*[-/年．.]\s*(\d{1,2})\s*[-/月．.]\s*(\d{1,2})\s*[日]?`)
var releaseDateLabelPattern = regexp.MustCompile(`(?i)(發行日期|发行日期|発売日|Release\s*Date|発売日期)`)

// ParseReleaseDate extracts the release date from a JavBus-style detail page.
// It searches for info rows whose header text matches a known date label and
// returns the first date found in YYYY-MM-DD form, or an empty string.
func ParseReleaseDate(htmlText string) string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return ""
	}

	var result string
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != "" {
			return
		}

		if node.Type == html.ElementNode && (node.Data == "p" || node.Data == "div" || node.Data == "span") {
			text := nodeText(node)
			if releaseDateLabelPattern.MatchString(text) {
				if matches := releaseDatePattern.FindStringSubmatch(text); len(matches) >= 4 {
					year := matches[1]
					month := matches[2]
					day := matches[3]
					result = fmt.Sprintf("%s-%02s-%02s", year, month, day)
					return
				}
			}
		}

		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

// ParseFilmData projects parsed metadata into the persisted crawl artifact
// shape used by organizer/subscription downstream.
func ParseFilmData(metadata Metadata, link string) FilmData {
	filmData := FilmData{
		Title:       strings.TrimSpace(metadata.Title),
		SourceLink:  strings.TrimSpace(link),
		Category:    append([]string(nil), metadata.Category...),
		Actress:     append([]string(nil), metadata.Actress...),
		ReleaseDate: strings.TrimSpace(metadata.ReleaseDate),
		Maker:       strings.TrimSpace(metadata.Maker),
		Label:       strings.TrimSpace(metadata.Label),
		Series:      strings.TrimSpace(metadata.Series),
	}
	if strings.TrimSpace(metadata.Img) != "" {
		filmData.CoverImage = strings.TrimSpace(metadata.Img)
	}
	return filmData
}

// ExtractAntiBlockURLs parses informational "anti block" links without mixing
// them into the main movie-link extraction path.
func ExtractAntiBlockURLs(htmlText string) []string {
	root, err := html.Parse(strings.NewReader(htmlText))
	if err != nil {
		return nil
	}
	result := make([]string, 0)
	seen := map[string]struct{}{}
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "div" && hasClass(node, "alert") && hasClass(node, "alert-info") {
			collectAntiBlockLinks(node, seen, &result)
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

func collectAntiBlockLinks(root *html.Node, seen map[string]struct{}, result *[]string) {
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "div" &&
			hasClass(node, "col-xs-12") &&
			hasClass(node, "col-md-6") &&
			hasClass(node, "col-lg-3") &&
			hasClass(node, "text-center") {
			if containsAnyIgnoreCase(nodeText(findFirstChild(node, "strong")), []string{"防屏蔽地址"}) {
				if linkNode := findFirstChild(node, "a"); linkNode != nil {
					if href := strings.TrimSpace(getAttr(linkNode, "href")); href != "" {
						if _, exists := seen[href]; !exists {
							seen[href] = struct{}{}
							*result = append(*result, href)
						}
					}
				}
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
}

func extractMetadataField(sources []string, patterns []*regexp.Regexp) string {
	for _, source := range sources {
		for _, pattern := range patterns {
			matches := pattern.FindStringSubmatch(source)
			if len(matches) >= 2 && strings.TrimSpace(matches[1]) != "" {
				return strings.TrimSpace(matches[1])
			}
		}
	}
	return ""
}

func collectElementTexts(root *html.Node, tag string) []string {
	result := make([]string, 0)
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if node.Type == html.ElementNode && strings.EqualFold(node.Data, tag) {
			text := strings.TrimSpace(nodeText(node))
			if text != "" {
				result = append(result, text)
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

func findFirstElementText(root *html.Node, tag string) string {
	var result string
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != "" {
			return
		}
		if node.Type == html.ElementNode && strings.EqualFold(node.Data, tag) {
			result = strings.TrimSpace(nodeText(node))
			return
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

func findMetaContentByProperty(root *html.Node, property string) string {
	var result string
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != "" {
			return
		}
		if node.Type == html.ElementNode && node.Data == "meta" && strings.EqualFold(getAttr(node, "property"), property) {
			result = strings.TrimSpace(getAttr(node, "content"))
			return
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

func findImageByAncestorClass(root *html.Node, ancestorTag string, className string) string {
	var result string
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != "" {
			return
		}
		if node.Type == html.ElementNode && node.Data == "img" {
			if hasAncestor(node, func(parent *html.Node) bool {
				return parent.Data == ancestorTag && hasClass(parent, className)
			}) {
				result = strings.TrimSpace(getAttr(node, "src"))
				return
			}
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(root)
	return result
}

func findFirstChild(root *html.Node, tag string) *html.Node {
	var result *html.Node
	var walk func(*html.Node)
	walk = func(node *html.Node) {
		if result != nil {
			return
		}
		if node.Type == html.ElementNode && strings.EqualFold(node.Data, tag) {
			result = node
			return
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	if root != nil {
		for child := root.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	return result
}

func hasAncestor(node *html.Node, predicate func(*html.Node) bool) bool {
	for parent := node.Parent; parent != nil; parent = parent.Parent {
		if predicate(parent) {
			return true
		}
	}
	return false
}

func hasClass(node *html.Node, className string) bool {
	classes := strings.Fields(strings.TrimSpace(getAttr(node, "class")))
	for _, item := range classes {
		if item == className {
			return true
		}
	}
	return false
}

func getAttr(node *html.Node, key string) string {
	for _, attr := range node.Attr {
		if strings.EqualFold(attr.Key, key) {
			return attr.Val
		}
	}
	return ""
}

func nodeText(node *html.Node) string {
	if node == nil {
		return ""
	}
	var builder strings.Builder
	var walk func(*html.Node)
	walk = func(current *html.Node) {
		if current.Type == html.TextNode {
			builder.WriteString(current.Data)
		}
		for child := current.FirstChild; child != nil; child = child.NextSibling {
			walk(child)
		}
	}
	walk(node)
	return strings.TrimSpace(builder.String())
}

func containsAnyIgnoreCase(value string, keywords []string) bool {
	normalized := strings.ToLower(value)
	for _, keyword := range keywords {
		if strings.Contains(normalized, strings.ToLower(keyword)) {
			return true
		}
	}
	return false
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}
